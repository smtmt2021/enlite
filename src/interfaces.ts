/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { EnliteSocket } from './enlite-impl';
import { Controller } from './controller';
import { Device } from './device';
import { Property } from './property';
import { ESV } from './esv';
import { CLASS } from './device-class';
import { WsunConfig, WsunId, WsunCache, NameOfClass } from '@smtmt2021/wsun-adaptor';

export interface FrameParams {
   tid?: number;
   seoj?: number;
   deoj?: number;
   esv: ESV;
   properties: Property[];
 }

export interface EnliteConfig {
  wsun: {
    adaptor: NameOfClass;
    id: WsunId;
    cache: WsunCache;
    config: WsunConfig;
  }
}

export interface SendToRemote {
  send(params: FrameParams, address: string): Promise<void>;
}

export interface TimeValue {
  time: Date;
  value: number;
}

export interface CreateDevice {
  createDevice(controller: Controller, deviceClass: CLASS, address: string, instanceCode?: number): Device | undefined;
}

export interface CreateSocket {
  createSocket(): EnliteSocket;
}

export interface DeviceConstructor {
  new(controller: Controller, address: string, deviceClass: CLASS, instanceCode?: number): Device;
}

export type EnliteLocal = CreateDevice & CreateSocket;
