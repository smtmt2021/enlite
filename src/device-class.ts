/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
export enum CLASS {
  NODE_PROFILE = 0x0ef0,
  CONTROLLER = 0x05ff,
  LOW_VOLTAGE_SMART_ELECTRIC_ENERGY_METER = 0x0288
}

export type ClassType = CLASS.NODE_PROFILE | CLASS.CONTROLLER | CLASS.LOW_VOLTAGE_SMART_ELECTRIC_ENERGY_METER;
