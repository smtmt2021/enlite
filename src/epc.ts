/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
export enum EPC {
   INSTANCE_LIST_NOTIFICATION = 0xd5,

   COEFFICIENT = 0xd3,
   EFFECTIVE_DIGITS = 0xd7,
   UNIT = 0xe1,
   INSTANTANEOUS_ELECTRIC_ENERGY = 0xe7,
   CUMULATIVE_AMOUNTS_OF_ENERGY_NORMAL = 0xea,
   CUMULATIVE_AMOUNTS_OF_ENERGY_REVERSE = 0xeb
 }
