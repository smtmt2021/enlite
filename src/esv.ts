/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
export enum ESV {
  GET = 0x62,
  GET_RES = 0x72,
  INF = 0x73,
  INFC = 0x74,
  INFC_RES = 0x7a
}
