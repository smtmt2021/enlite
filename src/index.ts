/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { Enlite } from './enlite';
import { EnliteImpl } from './enlite-impl';
export { Enlite };
export { SendToRemote as Controller } from './interfaces';
export { Device } from './device';
export { CLASS } from './device-class';
export { SmartElectricEnergyMeter } from './smart-electric-energy-meter';

export function EnliteGetInstance(): Enlite {
  return EnliteImpl.getInstance();
}
