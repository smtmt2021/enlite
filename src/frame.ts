/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { FrameParams } from './interfaces';
import { Property } from './property';
import { EPC } from './epc';
import { ESV } from './esv';
import { num2hex, dumpFrameParams, getResponseService } from './lib';

const EHD = 0x1081;

export class Frame {
  static create(params: FrameParams): Frame {
    if (!params.tid || !params.seoj || !params.deoj || !params.esv) {
      throw new Error(`Frame: params are not enough ${dumpFrameParams(params)}`);
    }
    let data = num2hex(EHD, 2);
    data += num2hex(params.tid, 2);
    data += num2hex(params.seoj, 3);
    data += num2hex(params.deoj, 3);
    data += num2hex(params.esv, 1);
    data += num2hex(params.properties.length, 1); // as OPC
    for (const property of params.properties) {
      data += num2hex(property.epc, 1) + num2hex(property.pdc, 1);
      if (property.edt) {
        for (const dt of property.edt.values()) {
          data += num2hex(dt, 1);
        }
      }
    }
    return new Frame(Buffer.from(data, 'hex'));
  }

  static confirmed(frame: Frame): FrameParams {
    const properties: Property[] = [];
    for (const property of frame.properties) {
      properties.push(Property.confirmed(property));
    }
    return {
      tid: frame.tid,
      seoj: frame.deoj,
      deoj: frame.seoj,
      esv: getResponseService(frame.esv),
      properties
    };
  }

  constructor(public readonly buffer: Buffer) {}

  get ehd(): number {
    return this.buffer.readUInt16BE(0);
  }

  set tid(tid: number) {
    this.buffer.writeUInt16BE(tid, 2);
  }

  get tid(): number {
    return this.buffer.readUInt16BE(2);
  }

  get seoj(): number {
    return this.buffer.readUIntBE(4, 3);
  }

  get deoj(): number {
    return this.buffer.readUIntBE(7, 3);
  }

  get esv(): ESV {
    return this.buffer.readUInt8(10) as ESV;
  }

  get opc(): number {
    return this.buffer.readUInt8(11);
  }

  get properties(): Property[] {
    const properties: Property[] = [];
    const buffer = this.buffer.subarray(12);
    let offset = 0;
    for (let i = 0; i < this.opc; ++i) {
      const epc = buffer[offset++] as EPC;
      const pdc = buffer[offset++];
      const edt = pdc ? buffer.subarray(offset, offset + pdc) : undefined;
      properties.push(new Property(epc, edt));
      offset += pdc;
    }
    return properties;
  }

  toString(): string {
    let result = 'ehd:' + num2hex(this.ehd, 2)
      + ' tid:' + num2hex(this.tid, 2)
      + ' seoj:' + num2hex(this.seoj, 3)
      + ' deoj:' + num2hex(this.deoj, 3)
      + ' esv:' + num2hex(this.esv, 1)
      + ' opc:' + num2hex(this.opc, 1)
      + ' ';
    if (this.opc) {
      for (const property of this.properties) {
        result += property.toString();
      }
    }
    return result;
  }
}
