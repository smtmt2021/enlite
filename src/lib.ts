/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { FrameParams, TimeValue } from './interfaces';
import { ESV } from './esv';

/**
 *  Convert number to hex-decimal string.
 *  @param n     to be converted.
 *  @param bytes number of bytes, e.g. 2 bytes means 4 digits.
 */
export function num2hex(n: number, bytes = 2): string {
  if (n > Number.MAX_SAFE_INTEGER) {
    throw new Error(`the specified number is too large, ${n}`);
  }
  if (bytes < 1 || bytes > 8) {
    throw new TypeError(`invalid size of bytes, $${bytes}`);
  }
  return n
    .toString(16)
    .toUpperCase()
    .padStart(bytes * 2, '0');
}

export function dumpFrameParams(params: FrameParams): string {
  let result = '';
  if (params.tid) {
    result += `tid:${num2hex(params.tid, 2)} `;
  }
  if (params.seoj) {
    result += `seoj:${num2hex(params.seoj, 3)} `;
  }
  if (params.deoj) {
    result += `deoj:${num2hex(params.deoj, 3)} `;
  }
  for (const property of params.properties) {
    result += property.toString();
  }
  return result;
}

export function getResponseService(esv: ESV): ESV {
  switch (esv) {
    case ESV.GET:
      return ESV.GET_RES;
    case ESV.INFC:
      return ESV.INFC_RES;
  }
  return esv;
}

export function edt2DateValue(edt?: Buffer): TimeValue | undefined {
  if (!edt || edt.length !== 11) {
    return;
  }
  const year = edt.readUInt16BE(0);
  const month = edt.readUInt8(2) - 1;
  const day = edt.readUInt8(3);
  const hour = edt.readUInt8(4);
  const minuite = edt.readUInt8(5);
  const second = edt.readUInt8(6);
  const value = edt.readUInt32BE(7);
  const time = new Date(year, month, day, hour, minuite, second, 0);
  return { time, value };
}

export function dateValue2Buffer(time: Date, value: number): Buffer {
  const buffer = Buffer.alloc(11);
  buffer.writeUInt16BE(time.getFullYear());
  buffer.writeUInt8(time.getMonth() + 1, 2);
  buffer.writeUInt8(time.getDate(), 3);
  buffer.writeUInt8(time.getHours(), 4);
  buffer.writeUInt8(time.getMinutes(), 5);
  buffer.writeUInt8(time.getSeconds(), 6);
  buffer.writeUInt32BE(value, 7);
  return buffer;
}
