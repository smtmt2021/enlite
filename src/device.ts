/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import promiseRetry from 'promise-retry';
import { WsunUdpTransmissionFailed } from '@smtmt2021/wsun-adaptor';
import { EventEmitter } from 'events';
import { Frame } from './frame';
import { Node } from './node';
import { Property } from './property';
import { SendToRemote } from './interfaces';
import { TIMEOUT_GET } from './constants';
import { CLASS } from './device-class';
import { ESV } from './esv';
import { EPC } from './epc';
import DEBUG from 'debug';
const debug = DEBUG('enlite/device');

export class GetResTimeout extends Error {
  constructor() {
    super('GetRes timeout');
    this.name = new.target.name;
  }
}

export abstract class Device extends Node {
  private event = new EventEmitter();

  constructor(
    protected controller: SendToRemote,
    public readonly address: string,
    deviceClass: CLASS,
    instanceCode = 1
  ) {
    super(deviceClass, instanceCode);
    debug('device: created', this.constructor.name);
  }

  async getValue(epc: EPC, byteLength: number): Promise<number | undefined> {
    debug('device.getValue');
    try {
      const requests = [new Property(epc)];
      const responce = await this.Get(requests);
      for (const property of responce) {
        if (property.epc === epc && property.edt?.length === byteLength) {
          return property.edt.readIntBE(0, byteLength);
        }
      }
    } catch (err) {
      console.error('device', err);
      throw err;
    }
  }

  protected Get(properties: Property[]): Promise<Property[]> {
    debug('device.Get');
    return promiseRetry((retry, count) =>
      this.getProperties(properties).catch(err => {
        if (err instanceof GetResTimeout || err instanceof WsunUdpTransmissionFailed) {
          console.info('enlite: Get(properties): retry', count);
          return retry(err);
        }
        throw err;
      }),
    {
      retries: 3,
      factor: 1
    });
  }

  protected getProperties(properties: Property[]): Promise<Property[]> {
    debug('device.getProperties');
    let timer: NodeJS.Timer;
    return new Promise<Property[]>((resolve, reject) => {
      this.event.once('getres', properties => resolve(properties));
      timer = setTimeout(() => reject(new GetResTimeout()), TIMEOUT_GET);
      this.controller.send({
        tid: this.getNextTid(),
        deoj: this.id,
        esv: ESV.GET,
        properties
      }, this.address).catch(err => reject(err));
    })
      .catch(err => {
        console.info('enlite: failed to get properties,', err);
        return Promise.reject(err);
      })
      .finally(() => {
        if (timer) {
          clearTimeout(timer);
        }
        this.event.removeAllListeners('getres');
      });
  }

  protected onGetRes(address: string, frame: Frame): Promise<void> {
    debug('device.onGetRes', frame.toString());
    if (this.tid !== frame.tid) {
      console.info('device.onGetRes: unexpected tid received', this.tid, frame.tid);
    } else {
      debug('device.onGetRes', frame.properties);
      this.event.emit('getres', frame.properties);
    }
    return Promise.resolve();
  }

  protected onInfC(address: string, frame: Frame): Promise<void> {
    debug('device.onInfC');
    console.info('enlite.onInfC:', frame.toString());
    const confirmed = Frame.confirmed(frame);
    return this.controller.send(confirmed, address);
  }
}
