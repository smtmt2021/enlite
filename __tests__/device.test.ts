/**
 * @license MIT License
 * @copyright KINOSHITA minoru, All Rights Reserved.
 */
import { Node } from '../src/node';
import { Device } from '../src/device';
import { Frame } from '../src/frame';
import { Property } from '../src/property';
import { FrameParams, SendToRemote } from '../src/interfaces';
import { getResponseService } from '../src/lib';
import { CLASS } from '../src/device-class';
import { ESV } from '../src/esv';
import { EPC } from '../src/epc';

class ControllerTest extends Node implements SendToRemote {
  send = jest.fn((params: FrameParams, address: string) => {
    params.seoj = this.id;
    return Promise.resolve();
  });
}

class DeviceTest extends Device {
  Get(properties: Property[]): Promise<Property[]> {
    return super.Get(properties);
  }
}

beforeAll(async () => {
  jest.setTimeout(20 * 1000);
});

test('device response to InfC', async () => {
  const controller = new ControllerTest(CLASS.CONTROLLER);
  const address = 'fe80:0000:0000:0000:0000:0000:0000:0001';
  const device = new DeviceTest(controller, address, CLASS.LOW_VOLTAGE_SMART_ELECTRIC_ENERGY_METER);
  const frame = new Frame(Buffer.from('1081029F02880105FF017402EA0B07E501121200000000FBD2EB0B07E5011212000000000009', 'hex'));

  await expect(controller.open()).resolves.toBeTruthy();
  await expect(device.open()).resolves.toBeTruthy();
  expect(device.onData(address, frame)).resolves.toBeUndefined();
  expect(controller.send.mock.calls.length).toBe(1);
  expect(controller.send.mock.calls[0][0]).toEqual({
    tid: 0x029f,
    seoj: 0x05ff01,
    deoj: 0x028801,
    esv: ESV.INFC_RES,
    properties: [
      new Property(EPC.CUMULATIVE_AMOUNTS_OF_ENERGY_NORMAL),
      new Property(EPC.CUMULATIVE_AMOUNTS_OF_ENERGY_REVERSE)
    ]
  });
  expect(controller.send.mock.calls[0][1]).toBe(address);
  await expect(device.close()).resolves.toBeTruthy();
  await expect(controller.close()).resolves.toBeTruthy();
});

test('Get', async () => {
  const controller = new ControllerTest(CLASS.CONTROLLER);
  const address = 'fe80:0000:0000:0000:0000:0000:0000:0003';
  const device = new DeviceTest(controller, address, CLASS.LOW_VOLTAGE_SMART_ELECTRIC_ENERGY_METER);
  const property = new Property(EPC.INSTANTANEOUS_ELECTRIC_ENERGY);
  const edt = Buffer.from('01234567', 'hex');

  await expect(controller.open()).resolves.toBeTruthy();
  await expect(device.open()).resolves.toBeTruthy();
  controller.send = jest.fn((params: FrameParams, address: string) => {
    [params.seoj, params.deoj] = [params.deoj, controller.id];
    params.esv = getResponseService(params.esv);
    params.properties = [new Property(EPC.INSTANTANEOUS_ELECTRIC_ENERGY, edt)];
    const frame = Frame.create(params);
    setTimeout(() => device.onData(address, frame), 100);
    return Promise.resolve();
  });
  await expect(device.Get([property])).resolves
    .toEqual([new Property(EPC.INSTANTANEOUS_ELECTRIC_ENERGY, edt)]);
  await expect(device.close()).resolves.toBeTruthy();
  await expect(controller.close()).resolves.toBeTruthy();
});

test('Get and timeout', async () => {
  const controller = new ControllerTest(CLASS.CONTROLLER);
  const address = 'fe80:0000:0000:0000:0000:0000:0000:0004';
  const device = new DeviceTest(controller, address, CLASS.LOW_VOLTAGE_SMART_ELECTRIC_ENERGY_METER);
  const property = new Property(EPC.INSTANTANEOUS_ELECTRIC_ENERGY);

  await expect(controller.open()).resolves.toBeTruthy();
  await expect(device.open()).resolves.toBeTruthy();
  await expect(device.Get([property])).rejects.toMatchSnapshot();
  await expect(device.close()).resolves.toBeTruthy();
  await expect(controller.close()).resolves.toBeTruthy();
});

test('Get and wrong tid, then timeout', async () => {
  const controller = new ControllerTest(CLASS.CONTROLLER);
  const address = 'fe80:0000:0000:0000:0000:0000:0000:0003';
  const device = new DeviceTest(controller, address, CLASS.LOW_VOLTAGE_SMART_ELECTRIC_ENERGY_METER);
  const property = new Property(EPC.INSTANTANEOUS_ELECTRIC_ENERGY);
  const edt = Buffer.from('01234567', 'hex');

  await expect(controller.open()).resolves.toBeTruthy();
  await expect(device.open()).resolves.toBeTruthy();
  controller.send = jest.fn((params: FrameParams, address: string) => {
    params.tid = 100;
    [params.seoj, params.deoj] = [params.deoj, controller.id];
    params.esv = getResponseService(params.esv);
    params.properties = [new Property(EPC.INSTANTANEOUS_ELECTRIC_ENERGY, edt)];
    const frame = Frame.create(params);
    setTimeout(() => device.onData(address, frame), 100);
    return Promise.resolve();
  });
  await expect(device.Get([property])).rejects.toMatchSnapshot();
  await expect(device.close()).resolves.toBeTruthy();
  await expect(controller.close()).resolves.toBeTruthy();
});

test('getValue(4)', async () => {
  const controller = new ControllerTest(CLASS.CONTROLLER);
  const address = 'fe80:0000:0000:0000:0000:0000:0000:0003';
  const device = new DeviceTest(controller, address, CLASS.LOW_VOLTAGE_SMART_ELECTRIC_ENERGY_METER);
  const property = new Property(EPC.INSTANTANEOUS_ELECTRIC_ENERGY);
  const edt = Buffer.from('12345678', 'hex');

  await expect(controller.open()).resolves.toBeTruthy();
  await expect(device.open()).resolves.toBeTruthy();
  controller.send = jest.fn((params: FrameParams, address: string) => {
    [params.seoj, params.deoj] = [params.deoj, controller.id];
    params.esv = getResponseService(params.esv);
    params.properties = [new Property(EPC.INSTANTANEOUS_ELECTRIC_ENERGY, edt)];
    const frame = Frame.create(params);
    setTimeout(() => device.onData(address, frame), 100);
    return Promise.resolve();
  });
  await expect(device.getValue(EPC.INSTANTANEOUS_ELECTRIC_ENERGY, 4)).resolves.toBe(0x12345678);
  await expect(device.close()).resolves.toBeTruthy();
  await expect(controller.close()).resolves.toBeTruthy();
});

test('getValue(1)', async () => {
  const controller = new ControllerTest(CLASS.CONTROLLER);
  const address = 'fe80:0000:0000:0000:0000:0000:0000:0003';
  const device = new DeviceTest(controller, address, CLASS.LOW_VOLTAGE_SMART_ELECTRIC_ENERGY_METER);
  const property = new Property(EPC.EFFECTIVE_DIGITS);
  const edt = Buffer.from('08', 'hex');

  await expect(controller.open()).resolves.toBeTruthy();
  await expect(device.open()).resolves.toBeTruthy();
  controller.send = jest.fn((params: FrameParams, address: string) => {
    [params.seoj, params.deoj] = [params.deoj, controller.id];
    params.esv = getResponseService(params.esv);
    params.properties = [new Property(EPC.EFFECTIVE_DIGITS, edt)];
    const frame = Frame.create(params);
    setTimeout(() => device.onData(address, frame), 100);
    return Promise.resolve();
  });
  await expect(device.getValue(EPC.EFFECTIVE_DIGITS, 1)).resolves.toBe(8);
  await expect(device.close()).resolves.toBeTruthy();
  await expect(controller.close()).resolves.toBeTruthy();
});
