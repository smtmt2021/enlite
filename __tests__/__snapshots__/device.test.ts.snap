// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get and timeout 1`] = `[GetResTimeout: GetRes timeout]`;

exports[`Get and wrong tid, then timeout 1`] = `[GetResTimeout: GetRes timeout]`;
